src/libstore/build/derivation-goal.cc

```diff
@@ -3121,6 +3121,20 @@ void DerivationGoal::registerOutputs()
             newInfo0.references = refs.second;
             if (refs.first)
                 newInfo0.references.insert(newInfo0.path);
+            if (scratchPath != newInfo0.path) {
+                // Also rewrite the output path
+                auto source = sinkToSource([&](Sink & nextSink) {
+                    StringSink sink;
+                    dumpPath(actualPath, sink);
+                    RewritingSink rsink2(oldHashPart, std::string(newInfo0.path.hashPart()), nextSink);
+                    rsink2((unsigned char *) sink.s->data(), sink.s->size());
+                    rsink2.flush();
+                });
+                Path tmpPath = actualPath + ".tmp";
+                restorePath(tmpPath, *source);
+                deletePath(actualPath);
+                movePath(tmpPath, actualPath);
```

Comment:
Maybe a call to `canonicalisePathMetaData()` is needed here. (See `rewriteOutputs()` which does so after doing a rewrite.)

Reply:
Indeed, thanks

Reply:
If we give `rewriteOutput` a references parameter above, I think we can deduplciate this?

1. `rewriteOutput(outputRewrites);`
2. calc hahes
3. ```
   if (scratchPath != newInfo0.path)
       rewriteOutput({scratchPath, newInfo0.path});
   ```

Reply:
Duh indeed. Thanks, should be fixed now